<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="base_frame" default="base_link"/>

  <arg name="mapping_registrator" default="gicp" description="ndt or gicp"/>
  <arg name="rviz" default="true"/>

  <arg name="lost_frame_max_acceleration" default="10.0"/>
  <arg name="local_map_num_keyframes" default="30"/>
  <arg name="dense_pointcloud_num_keyframes" default="20"/>
  <arg name="mapper_resolution" default="0.5"/>
  <arg name="mapper_max_iterations" default="500"/>
  <arg name="mapper_epsilon" default="0.001"/>
  <arg name="mapper_max_correspondence_distance" default="0.1"/>
  <arg name="lidar_calibration_max_frames" default="20"/>
  <arg name="calibration_eval_max_corr_distance" default="0.2"/>
  <arg name="solver_iterations" default="10"/>
  <arg name="calibration_skip_keyframes" default="10"/>

  <arg name="mapping_max_range" default="50.0"/>
  <arg name="max_calibration_range" default="40.0"/>

  <arg name="calibration_use_only_last_frames" default="true"/>
  <arg name="crop_z_calibration_pointclouds" default="true"/>
  <arg name="crop_z_calibration_pointclouds_value" default="4.0"/>

  <let name="calibration_camera_optical_link_frames" value="['']"/>

  <let name="calibration_lidar_frames" value="[
    pandar_qt_left,
    pandar_40p_right,
    pandar_qt_right,
    pandar_40p_front,
    pandar_qt_front,
    pandar_40p_rear,
    pandar_qt_rear]"/>

  <let name="mapping_lidar_frame" value="pandar_40p_left"/>
  <let name="mapping_pointcloud" value="/sensing/lidar/left_upper/outlier_filtered/pointcloud"/>
  <let name="detected_objects" value="/perception/object_recognition/detection/objects"/>

  <let name="calibration_camera_info_topics" value="['']"/>

  <let name="calibration_image_topics" value="[
    '']"/>

  <let
    name="calibration_pointcloud_topics"
    value="[
    /sensing/lidar/left_lower/outlier_filtered/pointcloud,
    /sensing/lidar/right_upper/outlier_filtered/pointcloud,
    /sensing/lidar/right_lower/outlier_filtered/pointcloud,
    /sensing/lidar/front_lower/outlier_filtered/pointcloud,
    /sensing/lidar/front_upper/outlier_filtered/pointcloud,
    /sensing/lidar/rear_lower/outlier_filtered/pointcloud,
    /sensing/lidar/rear_upper/outlier_filtered/pointcloud]"
  />

  <!-- mapping based calibrator -->
  <include file="$(find-pkg-share mapping_based_calibrator)/launch/calibrator.launch.xml">
    <arg name="ns" value=""/>
    <arg name="calibration_service_name" value="calibrate_lidar_lidar"/>

    <arg name="rviz" value="$(var rviz)"/>
    <arg name="base_frame" value="$(var base_frame)"/>

    <arg name="calibration_camera_optical_link_frames" value="$(var calibration_camera_optical_link_frames)"/>
    <arg name="calibration_lidar_frames" value="$(var calibration_lidar_frames)"/>
    <arg name="mapping_lidar_frame" value="$(var mapping_lidar_frame)"/>

    <arg name="mapping_pointcloud" value="$(var mapping_pointcloud)"/>
    <arg name="detected_objects" value="$(var detected_objects)"/>

    <arg name="calibration_camera_info_topics" value="$(var calibration_camera_info_topics)"/>
    <arg name="calibration_image_topics" value="$(var calibration_image_topics)"/>
    <arg name="calibration_pointcloud_topics" value="$(var calibration_pointcloud_topics)"/>

    <arg name="mapping_registrator" value="$(var mapping_registrator)"/>
    <arg name="lost_frame_max_acceleration" value="$(var lost_frame_max_acceleration)"/>
    <arg name="calibration_skip_keyframes" value="$(var calibration_skip_keyframes)"/>

    <arg name="local_map_num_keyframes" value="$(var local_map_num_keyframes)"/>
    <arg name="dense_pointcloud_num_keyframes" value="$(var dense_pointcloud_num_keyframes)"/>
    <arg name="mapper_resolution" value="$(var mapper_resolution)"/>
    <arg name="mapper_max_iterations" value="$(var mapper_max_iterations)"/>
    <arg name="mapper_epsilon" value="$(var mapper_epsilon)"/>
    <arg name="mapper_max_correspondence_distance" value="$(var mapper_max_correspondence_distance)"/>
    <arg name="lidar_calibration_max_frames" value="$(var lidar_calibration_max_frames)"/>
    <arg name="calibration_eval_max_corr_distance" value="$(var calibration_eval_max_corr_distance)"/>
    <arg name="solver_iterations" value="$(var solver_iterations)"/>
    <arg name="calibration_skip_keyframes" value="$(var calibration_skip_keyframes)"/>

    <arg name="mapping_max_range" value="$(var mapping_max_range)"/>
    <arg name="max_calibration_range" value="$(var max_calibration_range)"/>

    <arg name="calibration_use_only_last_frames" value="$(var calibration_use_only_last_frames)"/>
    <arg name="crop_z_calibration_pointclouds" value="$(var crop_z_calibration_pointclouds)"/>
    <arg name="crop_z_calibration_pointclouds_value" value="$(var crop_z_calibration_pointclouds_value)"/>
  </include>
</launch>
